core/trino-main/src/main/java/io/trino/execution/buffer/PageDeserializer.java

```diff
@@ -337,7 +337,7 @@ private void decompress()
                         blockSize,
                         sink.getSlice().byteArray(),
                         sink.getSlice().byteArrayOffset() + bytesPreserved,
-                        sink.getSlice().length());
+                        sink.getSlice().length() - bytesPreserved);
```

Comment:
Actually code below:
```
System.arraycopy(
                        source.getSlice().byteArray(),
                        source.getSlice().byteArrayOffset() + source.getPosition(),
                        sink.getSlice().byteArray(),
                        sink.getSlice().byteArrayOffset() + bytesPreserved,
                        blockSize);
```

seems broken too. Do we know that `blockSize` will always fit in the targetSlice given some of the data is taken by `bytesPreserved` @arhimondr ?

Reply:
```
int bufferSize = blockSizeInBytes
                        // to guarantee a single long can always be read entirely
                        + Long.BYTES;
                buffers[0] = new ReadBuffer(Slices.allocate(bufferSize))
```

The buffer size is allocated according to `blockSize` with an extra `8` bytes (one long) for carry-overs. 

The carry over is guaranteed to never be higher than `8` bytes as this is the most what can be requested to be read in one shot.

Reply:
👍